HW1/P5/P5-bfs.py

```diff
@@ -0,0 +1,117 @@
+            return (dist3, neighbors3, prev3)
+        outputRDD = inputRDD.flatMap(find_neighbor).reduceByKey(update_neighbor)
+        return outputRDD    
+
+    ## Find the path between the start and end nodes, by searching backwards in the [list of previous nodes]
+    def find_path(end, total_step, nodeRDD):
+        ## Set the end node
+        def set_end((node, (path, prev))):
+                if (node == end):
+                    return ((node, (total_step, prev)))
+                else:
+                    return ((node, (path, prev)))
+        ## First we select a random node from "prev" list
+        def update_path(inputRDD1, step):
+            def find_prev((node, (path, prev))):
+                prev_list = [(node, (path, prev))]
+                if path == step:
+                    pre = random.choice(prev)
+                    prev_list.append((pre, (step-1, [])))
+                return prev_list
+            ## Then we merge it to the current RDD with the updated "path" variable - indicating at which step the node is reached
+            outputRDD1 = inputRDD1.flatMap(find_prev).reduceByKey(lambda (x1,y1), (x2,y2): (max(x1,x2), y1+y2))
+            return outputRDD1
+
+        pathRDD = nodeRDD.map(lambda (node, (dist, neighbors, prev)): (node, (-1, prev)))
+        pathRDD2 = pathRDD.map(set_end)
+
+        #Loop through the steps by picking random previous nodes along the way, and output the picked path
+        for i in range(total_step, 0, -1):
+            pathRDD2 = update_path(pathRDD2, i)
+#         path_found = pathRDD2.map(lambda (k, (v1, v2)): (v1, k)).filter(lambda (k, v): k >= 0).sortByKey()
+        path_node = pathRDD2.filter(lambda (k, (v1, v2)): v1 >= 0).map(lambda (k, (v1, v2)): (v1, k)).sortByKey()
+        return path_node    
+    
+    ## Initialize root and count variables
+    nodeRDD = sourceRDD.map((set_root)).cache()
+    new_count = 1
+    visit_count = sc.accumulator(1)
+    print "Start from root:", root
+                                     
+    step = 0
+    while new_count > 0:
+        nodeRDD = update_node(nodeRDD, step, visit_count).cache()
+        new_count = nodeRDD.filter(lambda (k, v): v[0] == step+1).count()
+        visit_count += new_count
+        print "Step:", (step+1),"; New Nodes:", new_count,"; Total Nodes:", visit_count, "."
+        reach_end = nodeRDD.filter(lambda (k, v): k == end and len(v[2]) > 0).cache()
+        if reach_end.count() > 0:
+            print "Reach the end:", end
+            total_step = step+1
+            path_node = find_path(end, total_step, nodeRDD)
+            path_name_list = []
+            for i in path_node.collect():
+                path_name_list.append(lookupRDD.lookup(i[1]))
+            path_name = [x[0] for x in path_name_list]
+            print "The following path is found among all possible paths:"
+            print path_node.collect()
+            print path_name
+            return (path_node, path_name)
+        step += 1
```

Comment:
Similar comments from the bfs on 4 apply to the bfs on 5 -- it seems that, in your code, the accumulator functions as a large global variable that could have been used in a more useful way.

The rest is nice though and I appreciate all the documentation from 4 and 5!